### PR TITLE
Devel/rbtree

### PR DESCRIPTION
**Pull request by Werkov**

### CODE REVIEW

File: crash/types/rbtree.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        return parent
+
+def rbtree_postorder_for_each(root: gdb.Value) -> Iterable[gdb.Value]:
+    if not isinstance(root, gdb.Value):
+        raise InvalidArgumentError("root must be gdb.Value representing 'struct rb_root' or 'struct rb_root *' not {}"
+            .format(type(root).__name__))
+    if root.type == types.rb_root_type.pointer():
+        root = root.dereference()
+    elif root.type != types.rb_root_type:
+        raise ArgumentTypeError('root', root, types.rb_root_type)
```
Comment (jeffmahoney) [inline crash/types/rbtree.py]: This should be UnexpectedGDBTypeError now.

File: crash/types/rbtree.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    parent = rb_parent(node)
+    if (parent is not None and int(node) == int(parent['rb_left']) and
+            int(parent['rb_right']) != 0):
+        return rb_left_deepest_node(parent['rb_right'])
+    else:
+        return parent
+
+def rbtree_postorder_for_each(root: gdb.Value) -> Iterable[gdb.Value]:
+    if not isinstance(root, gdb.Value):
+        raise InvalidArgumentError("root must be gdb.Value representing 'struct rb_root' or 'struct rb_root *' not {}"
```
Comment (jeffmahoney) [inline crash/types/rbtree.py]: This should be ArgumentTypeError now

**Review by jeffmahoney — CHANGES_REQUESTED** [review]
There are a few nits, comments inline, but also (as you mentioned on IRC) since this is intended to be part of the API, we need documentation for it.  We also need test cases. If you can come up with simple cases like those in tests/test_list.py, that would be helpful.  Test cases that work with common in-kernel rbtrees would also be helpful to ensure that it still works correctly when the kernel is updated.

Finally, outside of the .gitignore changes, this should all be a single commit.

Thanks,

-Jeff

**Comment by Werkov** [comment]
No ready yet (only backing up the branch), will rebase later.

**Comment by Werkov** [comment]
Force pushed a new version:
- rebased onto current master,
- added unit and dump tests,
- squashed altogether.

Also, I noticed that similar API for linked list traversing says in the docs:

> def list_for_each
>         [...]
>         gdb.Value: The next node in the list.  The value is
>         of type ``struct list_head``.

However, the returned gdb.Value is of type `struct list_head *`, i.e. the docs isn't consistent with the implementation. The rbtree implementation honors the docs. Let me know which is the preferred way for these iterators (I expect the non-pointer variant).

**Comment by Werkov** [comment]
I've rebased the branch on the current master (resolved trouble with gdb update from 8.3 to 9.1) and added one more commit to rectify the type situation as laid out in my previous comment.
AFAICT, all the remarks should be resolved now.

**Pull request merged**